File: stats.c
/*
 * Lab 1: The Multi-Paradigm Tour -- C implementation.
 *
 * Compile: gcc -O2 -o stats_c stats.c
 * Run:     ./stats_c 4 8 15 16 23 42
 *
 * Complete the TODO section. See the assignment,
 * Part B, for the full shared contract (all three language versions
 * must match it exactly, including the tie-breaking mode rule).
 */

#include <stdio.h>
#include <stdlib.h>

int compare_ints(const void *a, const void *b) {
    int int_a = *(const int *)a;
    int int_b = *(const int *)b;
    return (int_a > int_b) - (int_a < int_b);
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        return 1;
    }

    int n = argc - 1;
    int *nums = malloc(n * sizeof(int));
    if (nums == NULL) {
        return 1;
    }
    for (int i = 0; i < n; i++) {
        nums[i] = atoi(argv[i + 1]);
    }

    /* TODO: compute mean, median, and mode from `nums` (length n).
     * - median: for even n, average the two middle values of the SORTED array.
     * - mode: most frequent value; on a tie, the SMALLEST tied value.
     * A sorted copy (via qsort + compare_ints) is useful for the median;
     * a simple O(n^2) frequency count is fine for mode at this input size.
     */

    double mean = 0.0;
    double median = 0.0;
    int mode = 0;

    /* Mean */
    if (n > 0) {
        long total = 0;
        for (int i = 0; i < n; i++) {
            total += nums[i];
        }
        mean = (double)total / n;
    }

    /* Median */
    int *sorted_nums = malloc(n * sizeof(int));
    if (sorted_nums == NULL) {
        free(nums);
        return 1;
    }
    for (int i = 0; i < n; i++) {
        sorted_nums[i] = nums[i];
    }
    qsort(sorted_nums, n, sizeof(int), compare_ints);

    if (n > 0) {
        int mid = n / 2;
        if (n % 2 == 0) {
            median = (sorted_nums[mid - 1] + sorted_nums[mid]) / 2.0;
        } else {
            median = sorted_nums[mid];
        }
    }

    /* Mode */
    /* O(n^2): for each number, count how many times it appears in nums.
     * Track the highest frequency seen and the smallest value that
     * achieves it, walking nums in ascending order so the
     * first value to reach a new max frequency is the smallest tied one. */
    int max_freq = 0;
    for (int i = 0; i < n; i++) {
        int candidate = sorted_nums[i];
        int freq = 0;
        for (int j = 0; j < n; j++) {
            if (nums[j] == candidate) {
                freq++;
            }
        }
        if (freq > max_freq) {
            max_freq = freq;
            mode = candidate;
        }
    }

    printf("Mean: %.2f\n", mean);
    printf("Median: %.2f\n", median);
    printf("Mode: %d\n", mode);

    free(nums);
    return 0;
}

File: stats.go
// Lab 1: The Multi-Paradigm Tour -- Go implementation.
//
// Compile: go build -o stats_go stats.go
// Run:     ./stats_go 4 8 15 16 23 42
//
// Complete the TODO section. See the assignment,
// Part B, for the full shared contract (all three language versions
// must match it exactly, including the tie-breaking mode rule).

package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
)

func computeStats(nums []int) (float64, float64, int) {
	// TODO: compute mean, median, and mode from nums.
	// - median: for an even count, average the two middle values of a SORTED copy.
	// - mode: most frequent value; on a tie, the SMALLEST tied value.
	// sort.Ints(sortedCopy) and a map[int]int frequency count will help.
	var mean float64
	var median float64
	var mode int

	n := len(nums)

	// mean
	if n > 0 {
		total := 0
		for _, v := range nums {
			total += v
		}
		mean = float64(total) / float64(n)
	}

	// median
	sortedCopy := make([]int, n)
	copy(sortedCopy, nums)
	sort.Ints(sortedCopy)

	if n > 0 {
		mid := n / 2
		if n%2 == 0 {
			median = float64(sortedCopy[mid-1]+sortedCopy[mid]) / 2.0
		} else {
			median = float64(sortedCopy[mid])
		}
	}

	// mode
	frequency := make(map[int]int)
	for _, v := range nums {
		frequency[v]++
	}

	if len(frequency) > 0 {
		maxFreq := 0
		for _, v := range sortedCopy { // ascending order, so ties resolve to smallest
			f := frequency[v]
			if f > maxFreq {
				maxFreq = f
				mode = v
			}
		}
	}
	return mean, median, mode
}

func main() {
	if len(os.Args) < 2 {
		os.Exit(1)
	}

	nums := make([]int, 0, len(os.Args)-1)
	for _, arg := range os.Args[1:] {
		n, err := strconv.Atoi(arg)
		if err != nil {
			os.Exit(1)
		}
		nums = append(nums, n)
	}

	mean, median, mode := computeStats(nums)
	fmt.Printf("Mean: %.2f\n", mean)
	fmt.Printf("Median: %.2f\n", median)
	fmt.Printf("Mode: %d\n", mode)
	_ = sort.Ints // keep import used even before TODO is filled in
	os.Exit(0)
}
